[AMDGPU] Relax lds dma waitcnt with no aliasing pair

kerbowa opened the pull request:
If we cannot find any lds DMA instruction that is aliased by some load from lds, we will still insert vmcnt(0). This is overly cautious since handling inter-thread dependences is normally managed by the memory model instead of the waitcnt pass, so this change updates the behavior to be more inline with how other types of memory events are handled.

kerbowa commented:
* **#131842** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/131842?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/131842?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

kerbowa commented:
I'm also debating on doing this by default instead of supplying a cl option.

cdevadas commented on llvm/test/CodeGen/AMDGPU/relax-lds-dma-waitcnt.ll:
```diff
@@ -0,0 +1,146 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
Can you pre-commit the test?

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -1748,7 +1753,7 @@ bool SIInsertWaitcnts::generateWaitcntInstBefore(MachineInstr &MI,
           }
         }
-        if (!FoundAliasingStore)
+        if (!FoundAliasingStore && !RelaxLDSDMA)
```
Is this still correct w.r.t. the check above?
```
if (Ptr && Memop->getAAInfo() && Memop->getAAInfo().Scope)
```
If that is false, I assume we want the wait anyway because we couldn't do alias analysis. Should we still add the conservative wait in that case?

kerbowa replied:
After including this check, I am just enabling this relaxation by default instead of as a cl option.

Pierre-vh replied:
Can you add a testcase for it too, please?

kerbowa replied:
The else case is tested in lds-dma-waitcnt.mir already.

kerbowa committed:
[AMDGPU] Relax lds dma waitcnt with no aliasing pair

If we cannot find any lds DMA instruction that is aliased by some load
from lds, we will still insert vmcnt(0). This is overly cautious since
handling inter-thread dependences is normally managed by the memory
model instead of the waitcnt pass, so this change updates the behavior
to be more inline with how other types of memory events are handled.

jayfoad commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -1768,13 +1767,12 @@ bool SIInsertWaitcnts::generateWaitcntInstBefore(MachineInstr &MI,
           const auto &LDSDMAStores = ScoreBrackets.getLDSDMAStores();
           for (unsigned I = 0, E = LDSDMAStores.size(); I != E; ++I) {
             if (MI.mayAlias(AA, *LDSDMAStores[I], true)) {
```
Nit: don't need the braces
